x-pack/solutions/chat/plugins/workchat-framework/server/framework/workflows/run_node.ts

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  WorkflowExecutionError,
+  type NodeFactoryBaseServices,
+  type NodeFactoryContext,
+  type ScopedRunnerRunNodeParams,
+} from '@kbn/wc-framework-types-server';
+import type { WorkflowRunnerInternalContext, ScopedNodeRunnerFn } from './types';
+import { createNoopNodeEventReporter, createNodeEventReporter, enterNode } from './utils';
+
+/**
+ * Returns a step runner already scoped to the given context.
+ */
+export const createNodeRunner = (params: {
+  internalContext: WorkflowRunnerInternalContext;
+}): ScopedNodeRunnerFn => {
+  return (args) => {
+    return runNode({
+      ...params,
+      ...args,
+    });
+  };
+};
+
+type RunStepParams = ScopedRunnerRunNodeParams & {
+  internalContext: WorkflowRunnerInternalContext;
+};
+
+const runNode = async ({
+  nodeDefinition,
+  state,
+  internalContext,
+}: RunStepParams): Promise<void> => {
+  // update the execution state - context is already a copy, it's safe to reference
+  internalContext.executionState = enterNode({
+    parent: internalContext.executionState,
+    nodeId: nodeDefinition.id,
+  });
+  const { nodeRegistry, eventHandler } = internalContext;
+
+  if (!nodeRegistry.has(nodeDefinition.type)) {
+    throw new WorkflowExecutionError(
+      `Node type [${nodeDefinition.type}] not found in registry`,
+      'nodeTypeNotFound',
+      { state: internalContext.executionState }
+    );
+  }
+
+  const nodeType = nodeRegistry.get(nodeDefinition.type);
+
+  const nodeServices = createBaseNodeServices({ internalContext });
+  // TODO: check / call nodeType.customServicesProvider if present
```

Comment:
Question: Just verifying, this TODO is meant for another PR?

Reply:
Correct, this code isn't "wired" to anything yet, this is still a work in progress